drivers/gpu/drm/rp1/rp1-vec/rp1_vec_hw.c

```diff
@@ -525,32 +574,48 @@ void rp1vec_hw_setup(struct rp1_vec *vec,
 
 void rp1vec_hw_update(struct rp1_vec *vec, dma_addr_t addr, u32 offset, u32 stride)
 {
+	unsigned long flags;
+
+	spin_lock_irqsave(&vec->hw_lock, flags);
+
 	/*
 	 * Update STRIDE, DMAH and DMAL only. When called after rp1vec_hw_setup(),
 	 * DMA starts immediately; if already running, the buffer will flip at
-	 * the next vertical sync event.
+	 * the next vertical sync event. In interlaced mode, we need to adjust
+	 * the address and stride to display only the current field, saving
+	 * the original address (so it can be flipped for subsequent fields).
 	 */
-	u64 a = addr + offset;
-
-	if (vec->fake_31khz) {
-		a += stride;
+	addr += offset;
+	vec->last_dma_addr = addr;
```

Comment:
Here, perhaps?

Reply:
It could skip one line `addr += offset;` but it's then needed to protect the atomicity of `last_dma_addr` and `last_stride` as well as the register writes.

Reply:
Moved